codex/stores/blockstore.nim

```diff
@@ -26,24 +28,32 @@ type
   BlockType* {.pure.} = enum
     Manifest, Block, Both
 
-  GetNext* = proc(): Future[?Cid] {.upraises: [], gcsafe, closure.}
-
-  BlocksIter* = ref object
-    finished*: bool
-    next*: GetNext
-
   BlockStore* = ref object of RootObj
 
-iterator items*(self: BlocksIter): Future[?Cid] =
-  while not self.finished:
-    yield self.next()
-
 method getBlock*(self: BlockStore, cid: Cid): Future[?!Block] {.base.} =
   ## Get a block from the blockstore
   ##
 
   raiseAssert("Not implemented!")
 
+method getBlock*(self: BlockStore, treeCid: Cid, index: Natural, merkleRoot: MultiHash): Future[?!Block] {.base.} =
+  ## Get a block by Cid of a merkle tree and an index of a leaf in a tree, validate inclusion using merkle root
+  ## 
+  
+  raiseAssert("Not implemented!")
+
+method getBlockAndProof*(self: BlockStore, treeCid: Cid, index: Natural): Future[?!(Block, MerkleProof)] {.base.} =
+  ## Get a block and associated inclusion proof by Cid of a merkle tree and an index of a leaf in a tree
+  ## 
+  
+  raiseAssert("Not implemented!")
+
+method getBlocks*(self: BlockStore, treeCid: Cid, leavesCount: Natural, merkleRoot: MultiHash): Future[?!AsyncIter[?!Block]] {.base.} =
```

Comment:
Maybe it makes sense to change the interface to take a `BlockAddress` everywhere?

Reply:
I added a generic method below, but I also left the individual methods for requesting either by `cid` or `(treeCid, index)` as well.
```
getBlock*(self: BlockStore, address: BlockAddress)
```